strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+                    )
+                prob_dist_val = np.real_if_close(
+                    np.sum(
+                        (
+                            np.array(self.weights)
+                            / np.sqrt(
+                                np.linalg.det(
+                                    2
+                                    * np.pi
+                                    * (self.covs[:, expind, :][:, :, expind].real + covmat)
+                                )
+                            )
+                        )
+                        * np.exp(-0.5 * exp_arg)
+                    )
+                )
+                prob_upbnd = np.real_if_close(
+                    np.sum(
+                        (
+                            ub_weights
+                            / np.sqrt(
+                                np.linalg.det(
+                                    2
+                                    * np.pi
+                                    * (
+                                        self.covs[ub_ind, :, :][:, expind, :][:, :, expind].real
+                                        + covmat
+                                    )
+                                )
+                            )
+                        )
+                        * np.exp(-0.5 * ub_exp_arg[ub_ind])
+                    )
+                )
+                vertical_sample = np.random.random(size=1) * prob_upbnd
+                if vertical_sample < prob_dist_val:
+                    drawn = True
+                    vals[i] = peak_sample
+        # The next line is a hack in that it only updates conditioned on the first samples value
+        # should still work if shots = 1
+        if len(indices) < len(self.active):
+            self.post_select_generaldyne(covmat, indices, vals[0])
+
+        # If all modes are measured, set them to vacuum
+        if len(indices) == len(self.active):
+            for i in indices:
+                self.loss(0, i)
+
+        return vals
+
+    def homodyne(self, n, shots=1, eps=0.0002):
+        r"""Performs an x-homodyne measurement on a mode, simulated by a generaldyne
+        onto a highly squeezed state.
+
+        Args:
+            n (int): mode to be measured
+            shots (int): how many measurements are performed
+            eps (int): squeezing of the measurement state
+
+        Returns:
+            array: homodyne outcome
+        """
+        covmat = self.hbar * np.diag(np.array([eps ** 2, 1.0 / eps ** 2])) / 2
+        return self.measure_dyne(covmat, [n], shots=shots)
+
+    def heterodyne(self, n, shots=1):
+        r"""Performs a heterodyne measurement on a mode, simulated by a generaldyne
+        onto a coherent state.
+
+        Args:
+            n (int): mode to be measured
+            shots (int): how many measurements are performed
+
+        Returns:
+            array: heterodyne outcome
+        """
+        covmat = self.hbar * np.eye(2) / 2
```

Comment:
If we now that `self.hbar` would not change in the backend, hardcoding in `covmat` might bring some slight performance advantage

Reply:
It seems safer to me to explicitly leave hbar everywhere in case someone wants to use the backend functions directly.